### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries/Linearity): move/add statements on linearity of L-series

### PR DESCRIPTION
**Pull request by MichaelStollBayreuth**
This adds a file `NumberTheory.LSeries.Linearity`, which contains statements on
* addition (moved from `NumberTheory.LSeries.Basic`)
* negation and
* scalar multiplication

of L-series, and corresponding statements for `LSeries.term`, `LSeriesHasSum` and `LSeriesSummable`.

See [this thread](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/L-series/near/424858837) on Zulip.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:66]:
```suggestion
lemma LSeriesSummable.neg {f : ℕ → ℂ} {s : ℂ} (hf : LSeriesSummable f s) :
```

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:72]:
```suggestion
lemma LSeriesSummable.neg_iff {f : ℕ → ℂ} {s : ℂ} :
```

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:94]:
```suggestion
lemma LSeriesHasSum.smul {f : ℕ → ℂ} (c : ℂ) {s a : ℂ} (hf : LSeriesHasSum f s a) :
```

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:98]:
```suggestion
lemma LSeriesSummable.smul {f : ℕ → ℂ} (c : ℂ) {s : ℂ} (hf : LSeriesSummable f s) :
```

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:111]:
```suggestion
lemma LSeries.smul (f : ℕ → ℂ) (c s : ℂ) : LSeries (c • f) s = c * LSeries f s := by
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
Upper-case types should be camelCased in theorem names, except in namespaces:

**Comment by MichaelStollBayreuth** [comment]
Changed names to be namespaced (and thus enabling dot notation) - thanks!

(I would plead for an exception to the lowerCamelCase rule for `LSeries...`, as the `L` is always upper case in informal math.)

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:20]: You can probably set f,g and some of a,b,c,s as implicit variables at the top and save having to create them each time.

Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:20]: I don't think this would be an improvement (see [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/'variable'.20-.20style.20question/near/425503092)).

**Comment by CBirkbeck** [comment]
> Changed names to be namespaced (and thus enabling dot notation) - thanks!
> 
> (I would plead for an exception to the lowerCamelCase rule for `LSeries...`, as the `L` is always upper case in informal math.)

LGTM! I also think the L should be upper case, but I'll let the maintainers decide.

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:24]:
```suggestion
  · simp only [term_of_ne_zero (Nat.succ_ne_zero _), Pi.add_apply, add_div]
```

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (loefflerd) [inline Mathlib/NumberTheory/LSeries/Linearity.lean:106]: Is `smul_apply` perhaps a better name?

**Review by loefflerd — APPROVED** [review]
I left a couple of small golfing comments. Otherwise looks good. 

(This PR does make me wonder whether we should be supporting `LSeries` valued in other spaces than `ℂ`, perhaps in a normed `ℂ`-vector space. This could come in useful, e.g. Iwasawa theorists like me are very fond of equivariant L-functions taking values in a group ring `ℂ[G]` where `G` is the Galois group of some finite extension. But it might have downsides in terms of making the API less convenient for the – undeniably most important – case of `ℂ` coefficients.)

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:106]: Maybe. But then it should also be `LSeries.add_apply` and `LSeries.neg_apply`, and one could have
```lean
lemma LSeries.smul (f : ℕ → ℂ) (c : ℂ) : LSeries (c • f) = c • LSeries f := ...
lemma LSeries.neg (f : ℕ → ℂ) : LSeries (-f) = -LSeries f := ...
```
This does not work for `add`, however, because of possibly differing convergence behaviour.

On the other hand, the other statements (`LSeries.term_smul`, `LSeriesHasSum.smul`, `LSeriesSummable.smul` etc.) all include the `s` argument, so the present version is reasonably consistent (and `LSeries.term_smul_apply` adds the `n` argument, not the `s` argument...).

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:106]: But `LSeriesSummable` and `LSeriesHasSum` are `Prop`s, while `LSeries` contains data, so the situation is not quite comparable. I would have expected `LSeries.smul` to be the map `ℂ → LSeries → LSeries` that defines the scalar multiplication operation (by analogy with `Nat.add` etc).

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:106]: PS: Possibly a better comparison is that there is `HasSum.const_smul` and `Summable.const_smul`, but there is no `tsum.const_smul`; there are no lemmas in the `tsum` namespace, because the output of `tsum` is a number, it doesn't have type `tsum`, so you can't dot-notate on it.

In fact this suggests (to me) that the naming convention `LSeries.foo` is misguided for all values of `foo` (not just `add`, `smul` etc), and they should all be `LSeries_foo` or indeed `lseries_foo`.

**Comment by MichaelStollBayreuth** [comment]
> (This PR does make me wonder whether we should be supporting `LSeries` valued in other spaces than `ℂ`, perhaps in a normed `ℂ`-vector space. This could come in useful, e.g. Iwasawa theorists like me are very fond of equivariant L-functions taking values in a group ring `ℂ[G]` where `G` is the Galois group of some finite extension. But it might have downsides in terms of making the API less convenient for the – undeniably most important – case of `ℂ` coefficients.)

I think it makes sense to try to get a reasonably complete set-up for complex-valued L-series first, and worry about generalizing it later. I don't expect a lot of problems with replacing `ℂ` as the target by a normed `ℂ`-vector space (or normed `ℂ`-algebra when talking about convolution).

File: Mathlib/NumberTheory/LSeries/Linearity.lean
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:106]: @Ruben-VandeVelde suggested the change from `LSeries_smul` to `LSeries.smul` (and the same for `neg`) a few days ago.
I don't have strong feelings either way, but I would like to have some kind of consensus here, so that we can move on...

Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/LSeries/Linearity.lean:106]: > In fact this suggests (to me) that the naming convention `LSeries.foo` is misguided for all values of `foo` (not just `add`, `smul` etc), and they should all be `LSeries_foo` or indeed `lseries_foo`.

Enabling dot notation is not the only reason to put stuff in namespaces, though. My understanding is that this also servers as an organising principle (e.g., in a currently active Zulip thread, it is dicsussed to introduce a `Discrete` namespace to serve as a home for statements on discrete Fourier transform etc., even though there won't be things of type `Discrete ...` to which they apply).

**Comment by loefflerd** [comment]
I'm happy with the current naming setup, and nobody else seems to have objected to it on the zulip thread, so let's get this in.

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**
